IntegrationTests/UITests/UITest.cs

```diff
@@ -0,0 +1,231 @@
+                    () =>
+                    {
+                        mainForm = new Form { Text = $"Test {typeof(T).Name}" };
+                        mainForm.Shown += (s, e) =>
+                        {
+                            showDialog(mainForm);
+                        };
+
+                        if (Application.MessageLoop)
+                        {
+                            mainForm.Show(owner: null);
+                        }
+                        else
+                        {
+                            Application.Run(mainForm);
+                        }
+                    },
+                    async (dialogForm) =>
+                    {
+                        try
+                        {
+                            await runTestAsync(dialogForm);
+                        }
+                        finally
+                        {
+                            CloseOrDispose(mainForm, $"{MethodBase.GetCurrentMethod().Name} main");
+                            mainForm = null;
+                        }
+                    });
+            }
+            finally
+            {
+                CloseOrDispose(mainForm, $"{MethodBase.GetCurrentMethod().Name} finally main");
+            }
+        }
+
+        private static void CloseOrDispose(Form form, string description)
+        {
+            try
+            {
+                ConfigureJoinableTaskFactoryAttribute.IgnoreExceptions = true;
+
+                if (form == null)
+                {
+                    LoggingService.Log($"{description} form is null");
+                    return;
+                }
+
+                var actionDescription = description + " form ";
+                try
+                {
+                    if (form.Visible)
+                    {
+                        actionDescription += "close";
+                        form.Close(); // also calls form.Dispose()
+                    }
+                    else
+                    {
+                        actionDescription += "dispose";
+                        form.Dispose();
+                    }
+
+                    LoggingService.Log($"{actionDescription}d");
+                }
+                catch (Exception ex)
+                {
+                    LoggingService.Log($"{actionDescription} threw {ex.Demystify()}");
+                    form.Dispose();
+                    LoggingService.Log($"{description} form disposed");
+                }
+            }
+            finally
+            {
+                // Wait for potential pending asynchronous tasks triggered by the form.
+                WaitForPendingOperations();
+                LoggingService.Log($"{description} form operations done");
+            }
+        }
+
+        private static void WaitForPendingOperations()
+        {
+            // Workaround for canceled operations: Process the message loop before and after the wait.
+            Application.DoEvents();
```

Comment:
😕 Why are we calling `Application.DoEvents`? This should not be necessary with the way JTF is configured for UI testing.

Reply:
Unfortunately, it was necessary. Refer to a7a3c6c4714a5f97cf61353c6323c9e27e8bca84.
I can double check this, i.e. try to remove these calls and let run the tests again.

Reply:
Without `Application.DoEvents`, two tests hang ([AppVeyor build](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/29908144)) and get killed by the test runner after the configured timeout.
```
1) Failed : GitExtensions.UITests.Script.ScriptRunnerTests.RunScript_with_arguments_with_s_option_with_RevisionGrid_with_selection_shall_succeed
Test exceeded Timeout value of 90000ms
2) Failed : GitExtensions.UITests.Script.ScriptRunnerTests.RunScript_with_arguments_with_s_option_with_RevisionGrid_without_selection_shall_display_error_and_return_false
Test exceeded Timeout value of 90000ms
```
The first of these tests must wait again for pending operations because the loading of FormBrowse had not completed yet. This wait timed out:
```
GitExtensions.UITests.Script.ScriptRunnerTests.RunScript_with_arguments_with_s_option_with_RevisionGrid_with_selection_shall_succeed
RunScript_with_arguments_with_s_option_with_RevisionGrid_with_selection_shall_succeed waits itself
WaitForPendingOperations cts canceled, ex System.OperationCanceledException: The operation was canceled.
   at void System.Threading.CancellationToken.ThrowOperationCanceledException()
   at void System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at async Task Microsoft.VisualStudio.Threading.ThreadingTools.WithCancellationSlow(Task task, bool continueOnCapturedContext, CancellationToken cancellationToken) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs:line 325
   at async Task Microsoft.VisualStudio.Threading.JoinableTaskCollection.JoinTillEmptyAsync(CancellationToken cancellationToken) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskCollection.cs:line 169
   at async Task GitUI.ThreadHelper.JoinPendingOperationsAsync(CancellationToken cancellationToken) in C:/projects/gitextensions/GitExtUtils/GitUI/ThreadHelper.cs:line 118
   at void Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread() in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTask.cs:line 914
   at void Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run(Func<Task> asyncMethod, JoinableTaskCreationOptions creationOptions, Delegate entrypointOverride) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs:line 554
   at void Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run(Func<Task> asyncMethod) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs:line 456
   at void CommonTestUtils.AsyncTestHelper.WaitForPendingOperations(CancellationToken cancellationToken) in C:/projects/gitextensions/UnitTests/CommonTestUtils/AsyncTestHelper.cs:line 67
   at void CommonTestUtils.AsyncTestHelper.WaitForPendingOperations(TimeSpan timeout) in C:/projects/gitextensions/UnitTests/CommonTestUtils/AsyncTestHelper.cs:line 51
WaitForPendingOperations cts canceled, ex System.OperationCanceledException: The operation was canceled.
   at void System.Threading.CancellationToken.ThrowOperationCanceledException()
   at void System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at async Task Microsoft.VisualStudio.Threading.ThreadingTools.WithCancellationSlow(Task task, bool continueOnCapturedContext, CancellationToken cancellationToken) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs:line 325
   at async Task Microsoft.VisualStudio.Threading.JoinableTaskCollection.JoinTillEmptyAsync(CancellationToken cancellationToken) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskCollection.cs:line 169
   at async Task GitUI.ThreadHelper.JoinPendingOperationsAsync(CancellationToken cancellationToken) in C:/projects/gitextensions/GitExtUtils/GitUI/ThreadHelper.cs:line 118
   at void Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread() in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTask.cs:line 914
   at void Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run(Func<Task> asyncMethod, JoinableTaskCreationOptions creationOptions, Delegate entrypointOverride) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs:line 554
   at void Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run(Func<Task> asyncMethod) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs:line 456
   at void CommonTestUtils.AsyncTestHelper.WaitForPendingOperations(CancellationToken cancellationToken) in C:/projects/gitextensions/UnitTests/CommonTestUtils/AsyncTestHelper.cs:line 67
   at void CommonTestUtils.AsyncTestHelper.WaitForPendingOperations(TimeSpan timeout) in C:/projects/gitextensions/UnitTests/CommonTestUtils/AsyncTestHelper.cs:line 51
HandleApplicationThreadException ignoring: System.Threading.ThreadAbortException: Thread was being aborted.
   at object RuntimeMethodHandle.InvokeMethod(object target, object[] arguments, Signature sig, bool constructor)
   at object System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(object obj, object[] parameters, object[] arguments)
   at object System.Reflection.RuntimeMethodInfo.UnsafeInvoke(object obj, BindingFlags invokeAttr, Binder binder, object[] parameters, CultureInfo culture)
   at object Delegate.DynamicInvokeImpl(object[] args)
   at object Delegate.DynamicInvoke(object[] args)
   at void System.Windows.Forms.Control.InvokeMarshaledCallbackDo(ThreadMethodEntry tme)
   at void System.Windows.Forms.Control.InvokeMarshaledCallbackHelper(object obj)
   at void System.Windows.Forms.Control.InvokeMarshaledCallback(ThreadMethodEntry tme)
   at void System.Windows.Forms.Control.InvokeMarshaledCallbacks()
HandleApplicationThreadException ignoring: System.Threading.ThreadAbortException: Thread was being aborted.
   at void System.Windows.Forms.Control.InvokeMarshaledCallbacks()
   at void System.Windows.Forms.Control.WndProc(ref Message m)
   at void System.Windows.Forms.Control+ControlNativeWindow.OnMessage(ref Message m)
   at void System.Windows.Forms.Control+ControlNativeWindow.WndProc(ref Message m)
   at IntPtr System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, int msg, IntPtr wparam, IntPtr lparam)
RunForm exception: System.Threading.ThreadAbortException: Thread was being aborted.
   at IntPtr System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, int msg, IntPtr wparam, IntPtr lparam)
   at IntPtr System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(ref MSG msg)
   at bool System.Windows.Forms.Application+ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, int reason, int pvLoopData)
   at void System.Windows.Forms.Application+ThreadContext.RunMessageLoopInner(int reason, ApplicationContext context)
   at void System.Windows.Forms.Application+ThreadContext.RunMessageLoop(int reason, ApplicationContext context)
   at void System.Windows.Forms.Application.Run(Form mainForm)
   at bool GitUI.GitUICommands.StartBrowseDialog(IWin32Window owner, string filter, ObjectId selectedCommit) in C:/projects/gitextensions/GitUI/GitUICommands.cs:line 1153
   at void GitExtensions.UITests.Script.ScriptRunnerTests.RunFormTest(Action<FormBrowse> testDriver)+() => { } in C:/projects/gitextensions/IntegrationTests/UITests/Script/ScriptRunnerTests.cs:line 210
   at void GitExtensions.UITests.UITest.RunForm<T>(Action showForm, Func<T, Task> runTestAsync) in C:/projects/gitextensions/IntegrationTests/UITests/UITest.cs:line 78
AfterTest cts cancelled: System.OperationCanceledException: The operation was canceled.
   at void System.Threading.CancellationToken.ThrowOperationCanceledException()
   at void System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at async Task Microsoft.VisualStudio.Threading.ThreadingTools.WithCancellationSlow(Task task, bool continueOnCapturedContext, CancellationToken cancellationToken) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs:line 325
   at async Task Microsoft.VisualStudio.Threading.JoinableTaskCollection.JoinTillEmptyAsync(CancellationToken cancellationToken) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskCollection.cs:line 169
   at async Task GitUI.ThreadHelper.JoinPendingOperationsAsync(CancellationToken cancellationToken) in C:/projects/gitextensions/GitExtUtils/GitUI/ThreadHelper.cs:line 118
   at void Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread() in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTask.cs:line 914
   at void Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run(Func<Task> asyncMethod, JoinableTaskCreationOptions creationOptions, Delegate entrypointOverride) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs:line 554
   at void Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run(Func<Task> asyncMethod) in E:/A/_work/398/s/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs:line 456
   at void CommonTestUtils.ConfigureJoinableTaskFactoryAttribute.AfterTest(ITest test) in C:/projects/gitextensions/UnitTests/CommonTestUtils/ConfigureJoinableTaskFactoryAttribute.cs:line 93
```